File: Extensions/Tessa.Extensions.Default.Client/Workflow/KrPermissions/KrTokenToTaskHistoryViewUIExtension.cs
using System.Threading.Tasks;
using Tessa.Cards;
using Tessa.Cards.TypeSettings;
using Tessa.Extensions.Default.Shared.Cards;
using Tessa.Extensions.Default.Shared.Workflow.KrProcess;
using Tessa.Platform.Storage;
using Tessa.UI.Cards;
using Tessa.UI.Cards.Controls;
using Tessa.Views;
using Tessa.Views.Metadata;
using Tessa.Views.Metadata.Criteria;

namespace Tessa.Extensions.Default.Client.Workflow.KrPermissions
{
    /// <summary>
    /// Устанавливает в параметре <see cref="CardTypeExtensionSettings.TokenParameterAlias"/> представления, указанного по ключу <see cref="CardTypeExtensionSettings.ViewControlAlias"/>, информацию о токене безопасности <see cref="KrToken"/>, выданного для карточки.
    /// </summary>
    public sealed class KrTokenToTaskHistoryViewUIExtension :
        CardUIExtension
    {
        #region Base Overrides

        /// <inheritdoc/>
        public override async Task Initializing(ICardUIExtensionContext context)
        {
            var result = await CardHelper
                .ExecuteTypeExtensionsAsync(
                    DefaultCardTypeExtensionTypes.MakeViewTaskHistory,
                    context.Card,
                    context.Model.CardMetadata,
                    ExecuteInitializingAsync,
                    context,
                    cancellationToken: context.CancellationToken);

            context.ValidationResult.Add(result);
        }

        #endregion

        #region Type extension methods

        private static Task ExecuteInitializingAsync(ITypeExtensionContext typeContext)
        {
            if (typeContext.ExternalContext is not ICardUIExtensionContext context)
            {
                return Task.CompletedTask;
            }

            var settings = typeContext.Settings;
            var tokenParameterAlias = settings?.TryGet<string>(CardTypeExtensionSettings.TokenParameterAlias);
            if (string.IsNullOrWhiteSpace(tokenParameterAlias))
            {
                return Task.CompletedTask;
            }

            var viewControlAlias = settings.TryGet<string>(CardTypeExtensionSettings.ViewControlAlias);

            // do not use closure for context/typeContext
            context.Model.ControlInitializers.Add((control, m, r, ct) =>
            {
                if (control is CardViewControlViewModel { ViewMetadata: { } viewMetadata, Parameters: { } parameters } viewControl
                    && viewControl.Name == viewControlAlias
                    && viewMetadata.Parameters.IsDefinedByName(tokenParameterAlias)
                    && KrToken.TryGet(m.Card.Info) is { } token)
                {
                    parameters.Add(new RequestParameter(tokenParameterAlias).Add(EqualsToCriteriaOperator.Instance, token.ToTypedJson()));
                }

                return ValueTask.CompletedTask;
            });

            return Task.CompletedTask;
        }

        #endregion
    }
}

File: WebClient SDK/src/default/workflow/krPermissions/krTokenToTaskHistoryViewUIExtension.ts
import { extension } from '@tessa/application';
import { tryGetFromSettings } from 'tessa/ui';
import { CardUIExtension, ICardUIExtensionContext } from 'tessa/ui/cards';
import { ViewControlViewModel } from 'tessa/ui/cards/controls';
import { KrToken } from 'tessa/workflow';
import {
  CardHelper,
  ICardTypeExtensionContext,
  CardTypeExtensionTypes,
  ViewRequestParameterBuilder,
  ViewCriteriaOperators,
  CardTypeExtensionSettings
} from '@tessa/platform';
import { StringHelper, TypedJsonConverter } from '@tessa/core';

/**
 * Устанавливает в параметре {@link CardTypeExtensionSettings.TokenParameterAlias} представления, указанного по ключу {@link CardTypeExtensionSettings.ViewControlAlias},
 * информацию о токене безопасности {@link KrToken}, выданного для карточки.
 */
@extension({ name: 'KrTokenToTaskHistoryViewUIExtension' })
export class KrTokenToTaskHistoryViewUIExtension extends CardUIExtension {
  //#region base overrides

  public async initializing(context: ICardUIExtensionContext): Promise<void> {
    const result = await CardHelper.executeTypeExtensions(
      CardTypeExtensionTypes.MakeViewTaskHistory,
      context.card,
      context.model.generalMetadata,
      KrTokenToTaskHistoryViewUIExtension.executeInitializingAction,
      context
    );

    context.validationResult.add(result);
  }

  //#endregion

  //#region private methods

  private static async executeInitializingAction(
    typeContext: ICardTypeExtensionContext
  ): Promise<void> {
    const context = typeContext.externalContext as ICardUIExtensionContext;
    if (!context) {
      return;
    }

    const settings = typeContext.settings;
    const viewControlAlias = tryGetFromSettings<string>(
      settings,
      CardTypeExtensionSettings.ViewControlAlias
    );

    if (!viewControlAlias) {
      return;
    }

    context.model.controlInitializers.push(async control => {
      const viewControl = control as ViewControlViewModel;

      if (!viewControl) {
        return;
      }

      if (viewControl.name !== viewControlAlias) {
        return;
      }

      const token = KrToken.tryGet(context.card.info);
      if (!token) {
        return;
      }

      const tokenParameterAlias = tryGetFromSettings<string>(
        settings,
        CardTypeExtensionSettings.TokenParameterAlias
      );
      if (StringHelper.isNullOrWhiteSpace(tokenParameterAlias)) {
        return;
      }

      const tokenMetadata = viewControl.viewMetadata?.parameters.tryGet(tokenParameterAlias);
      if (!tokenMetadata) {
        return;
      }

      const tokenString = TypedJsonConverter.serialize(token.getStorage());
      const tokenParameter = new ViewRequestParameterBuilder()
        .withMetadata(tokenMetadata)
        .addCriteria(ViewCriteriaOperators.EqualsTo, tokenString, tokenString)
        .asRequestParameter();
      viewControl.parameters.addParameters(tokenParameter);
    });
  }

  //#endregion
}
